plugins/SpectrumAnalyzer/SaWaterfallView.cpp

```diff
@@ -0,0 +1,232 @@
+										  displayBottom,
+										  Qt::IgnoreAspectRatio,
+										  Qt::SmoothTransformation));
+	}
+	else
+	{
+		painter.fillRect(displayLeft, displayTop, displayWidth, displayBottom, QColor(0,0,0));
+	}
+	lock.unlock();
+
+	// always draw the outline
+	painter.setPen(QPen(m_controls->m_colorGrid, 2, Qt::SolidLine, Qt::RoundCap, Qt::BevelJoin));
+	painter.drawRoundedRect(displayLeft, displayTop, displayWidth, displayBottom, 2.0, 2.0);
+
+	#ifdef SA_DEBUG
+		// display what FPS would be achieved if waterfall ran in a loop
+		start_time = std::chrono::high_resolution_clock::now().time_since_epoch().count() - start_time;
+		painter.setPen(QPen(m_controls->m_colorLabels, 1, Qt::SolidLine, Qt::RoundCap, Qt::BevelJoin));
+		painter.drawText(displayRight -100, 10, 100, 16, Qt::AlignLeft,
+						 QString(std::string("Max FPS: " + std::to_string(1000000000.0 / start_time)).c_str()));
+	#endif
+}
+
+
+// Convert time value to Y coordinate for display of given height.
+float SaWaterfallView::timeToYPixel(float time, int height)
+{
+	float pixels_per_line = (float)height / m_processor->m_waterfallHeight;
+	float seconds_per_line = ((float)m_processor->m_inBlockSize / m_processor->getSampleRate());
+
+	return pixels_per_line * time / seconds_per_line;
+}
+
+
+// Generate labels for linear time scale.
+std::vector<std::pair<float, std::string>> SaWaterfallView::makeTimeTics()
+{
+	std::vector<std::pair<float, std::string>> result;
+	float i;
+
+	// upper limit defined by number of lines * time per line
+	float limit = m_processor->m_waterfallHeight * ((float)m_processor->m_inBlockSize / m_processor->getSampleRate());
+
+	// set increment so that about 8 tics are generated
+	float increment = std::round(10 * limit / 7) / 10;
+
+	// NOTE: labels positions are rounded to match the (rounded) label value
+	for (i = 0; i <= limit; i += increment)
+	{
+		if (i < 10)
+		{
+			result.push_back(std::pair<float, std::string>(std::round(i * 10) / 10,
+							 std::to_string(std::round(i * 10) / 10).substr(0, 3)));
+		}
+		else
+		{
+			result.push_back(std::pair<float, std::string>(std::round(i),
+							 std::to_string(std::round(i)).substr(0, 2)));
+		}
+	}
+	return result;
+}
+
+
+// Periodically trigger repaint and check if the widget is visible.
+// If it is not, stop drawing and inform the processor.
+void SaWaterfallView::periodicUpdate()
+{
+	m_processor->setWaterfallActive(isVisible());
+	if (isVisible()) {update();}
+}
+
+
+// Adjust window size and widget visibility when waterfall is enabled or disabbled.
+void SaWaterfallView::updateVisibility()
+{
+	// get container of the control dialog to be resized if needed
+	QWidget *subWindow = m_controlDialog->parentWidget();
+
+
+	if (m_controls->m_waterfallModel.value())
+	{
+		// clear old data before showing the waterfall
+		QMutexLocker lock(&m_processor->m_dataAccess);
+		std::fill(m_processor->m_history.begin(), m_processor->m_history.end(), 0);
```

Comment:
This would just change the waterfall view from very simple (display image, done) to somewhat complicated -- it would need to adjust scaling of the image based on currently displayed number of lines, make sure the scaling changes are pixel-perfect so that there is no distortion / pulsing between subsequent frames and then make sure to redraw the remaining part of the background with black color.. For some small performance benefit that would disappear after a few seconds, making the performance inconsistent and potentially confusing.

As for the conceptual part -- the processor could probably do some visibility check as well, but it would be in a polling fashion. So this seems to me as a functionally better solution -- dump the history only if and when needed.

But anyway -- if we want the "realtime" solution without locking, the entire history building stuff will probably have to be moved from the processor to waterfall view. Then it will simply be cleaning "its own" history, so the conceptual problem goes away.

Reply:
I was not thinking about "scaling" the image, but I just meant you could restrict this index in order to draw only on a part of the image, and keep the rest of the image black. The image itself would not be resized, it would just contain a few black lines.

Good point with the "realtime" argument. I'm fine with either solution.

Reply:
The waterfall display just displays the entire stored raw image, I can't simply say "stop here and pretend the remaining values are zero". If I stop after some number of lines, the image will have wrong aspect ratio (literally just one line at the beginning) when scaled to the full size of the waterfall widget.
So it would have to be vertically resized to `(current_number_of_lines / max_number_of_lines) * widget height`, positioned on the top and the rest would have to be filled using a black rectangle, otherwise it would just stay the default gray background color. And if Qt happens to use integers for the scaling parameter, the exact required height may be rounded down, so any sharp edges would probably jump around (even more than they already do). I'm not saying it's not possible, but I don't see enough benefit for such complications.

Reply:
Oh, OK, I thought that was doable. I think we can resolve this one, too.